add Modbus Sunspec compatibility to opendtu-onbattery

AloisKlingler opened the pull request:
this implementation is (maybe only partially) modbus-sunspec compatibility to simulate a sunspec energy-meter.

to have calculation (especially self-consumption) correct for e.g. fronius solar web you need to either place a 300€ fronius smart meter before each "foreign" inverter.

With this PR it is possible to configure OpenDTU with it's IP-address as a "Smartmeter" on Fronius Gen24 or Fronius Symo inverter. so the Fronius Inverter is able to poll current production and overall production continuously like a from a real fronius-smartmeter.

necessary configuration on OpenDTU:
![image](https://github.com/user-attachments/assets/3d7b601d-9cb1-42c9-8ab7-2210ae29206b)

configuration on an Gen24 Inverter:
![image](https://github.com/user-attachments/assets/669f9f3a-50c9-4c5f-a229-aa945b63c72c)

visualisation on an Gen24 inverter:
![image](https://github.com/user-attachments/assets/1945a61b-fec7-4fa4-9a4e-19d0cdb30f54)

visualisation of the additional "Smartmeter" in Fronius Solarweb as "Additional AC Source":
![image](https://github.com/user-attachments/assets/122e5b52-8d8d-4d91-957e-dd5d4340f942)

visualisation of the "Smartmeter" data in Solarweb is available as common curve like other inverters are:
![image](https://github.com/user-attachments/assets/0070c442-d9bb-465a-95cc-5a4b8a5f25e9)


for my OpenDTU (and also some other users of photovoltaikforum.com) this implementation is running now over a year (more or less) flawless. 

to the PR: the tree conflicting files webapp_dist/* need to be resolved on merge. other conflicts I could solve.

ArekKubacki committed:
Add Modbus

ArekKubacki committed:
Added Modbus

ArekKubacki committed:
Merge remote-tracking branch 'origin/master'

ArekKubacki committed:
Checking that the measurement does not exceed the max Power

ArekKubacki committed:
Modbus BugFix

ArekKubacki committed:
Modbus BugFix

ArekKubacki committed:
Added WatchDog to Modbus

ArekKubacki committed:
Update from master

ArekKubacki committed:
Merge branch 'master' into master

ArekKubacki committed:
Revert "Merge branch 'master' into v1"

This reverts commit 5617240ce1fb15a3796d1ed558cc89f9ba72d8d5.

ArekKubacki committed:
Merge branch 'tbnobody-master'

ArekKubacki committed:
Edit platformio.ini

ArekKubacki committed:
Fixed Temperature, Frequency etc.

ArekKubacki committed:
Merge remote-tracking branch 'origin/master'

ArekKubacki committed:
Modbus fixed

ArekKubacki committed:
Fixed Temp, Freq, AC

ArekKubacki committed:
Merge remote-tracking branch 'origin/master'

ArekKubacki committed:
Merge

ArekKubacki committed:
Merge

ArekKubacki committed:
Modbus Added

ArekKubacki committed:
ModbusDtu h file fixed

ArekKubacki committed:
Fixed Modbus Library

bn-aklingler committed:
simulating a Fronius Smart Meter with openDTU. currently only feeding "Total Power" and "Total Yield" into registers.

bn-aklingler committed:
if only one inverter is connected, use its measurement data.

AloisKlingler committed:
Merge pull request #1 from tbnobody/master

merge 23.6.26

bn-aklingler committed:
adding totals as webapi live
prevent sending 0 as "e_energy_exported" to avoid peaks and gaps in solarweb statistic

bn-aklingler committed:
Merge branch 'master' of https://github.com/AloisKlingler/OpenDTU-FroniusSM-MB

AloisKlingler committed:
Merge pull request #2 from tbnobody/master

merge v23.6.28

bn-aklingler committed:
disable wifi powersave mode

AloisKlingler committed:
Merge pull request #3 from tbnobody/master

merge v23.8.3

bn-aklingler committed:
single phase inverters: use variable to fill them to phase 1 to smartmeter

bn-aklingler committed:
fixing still empty values on HMS inverters

AloisKlingler committed:
Merge pull request #4 from tbnobody/master

merge v23.8.8

bn-aklingler committed:
next try to avoid sending 0-values to modbus-register.

bn-aklingler committed:
missed some changes in last commit

AloisKlingler committed:
Merge pull request #5 from tbnobody/master

merge v23.8.28

AloisKlingler committed:
Merge pull request #6 from tbnobody/master

merge v23.9.11

AloisKlingler committed:
Merge pull request #7 from tbnobody/master

merge v23.9.13

bn-aklingler committed:
no more additional smartmeter data to increase reliability

AloisKlingler committed:
Merge pull request #8 from tbnobody/master

merge v23.9.18

AloisKlingler committed:
Merge pull request #9 from tbnobody/master

merge v23.10.9

AloisKlingler committed:
Merge pull request #10 from tbnobody/master

merge v23.11.16

bn-aklingler committed:
init variable on starting

bn-aklingler committed:
adding configuration element in settings -> network settings to enable / disable smartmeter simulation

bn-aklingler committed:
Revert "adding configuration element in settings -> network settings to enable / disable smartmeter simulation"

This reverts commit 1f77c38db212adc7245435e18a9b0499754bef2f.

bn-aklingler committed:
Revert "init variable on starting"

This reverts commit b95041ee2ccc804752f103520f26fcce9c651465.

bn-aklingler committed:
Revert "no more additional smartmeter data to increase reliability"

This reverts commit 1aaa4c27dc00583ca5f72ab3baef5dc89d13c2b9.

bn-aklingler committed:
Revert "missed some changes in last commit"

This reverts commit 2f1bb0def1f8ca60eef4e3bd65a04333d2166c81.

bn-aklingler committed:
Merge branch 'master' of https://github.com/tbnobody/OpenDTU into tbnobody-master

bn-aklingler committed:
Merge branch 'tbnobody-master'

bn-aklingler committed:
resolve mergeconflicts v24.6.29
remove watchdog
make simulated smartmeter serial dependent of DTU serial

bn-aklingler committed:
fix cpplint failing checks

bn-aklingler committed:
migrate Modbus to TaskScheduler
switch to emelianov modbus repository

bn-aklingler committed:
fix wrong taskname for modbustask

bn-aklingler committed:
v24.8.1-0.2.2 based on v24.8.1

bn-aklingler committed:
Merge branch 'master' of https://github.com/tbnobody/OpenDTU

bn-aklingler committed:
bring back detailled smartmeter data if only one hoymiles inverter is configured
start modbus task only if necessary

bn-aklingler committed:
wrong modbus register leaded to wrong Reactive Power in smartmeter data

AloisKlingler committed:
Update settings.json

AloisKlingler committed:
Update settings.json

AloisKlingler committed:
Update settings.json

bn-aklingler committed:
rectify settings.json

AloisKlingler committed:
Update main.cpp

bn-aklingler committed:
extract strings to have a better possibility to configure dynamic sunspec vendor data

AloisKlingler committed:
Merge pull request #15 from tbnobody/master

merge v24.8.5

AloisKlingler commented:
> thanks for your effort.
> 
> another thing, it would be good, if your changes would be based on helgeerbe:development or :master, too.
> 
> right now, your changes are based on a fork of the upstream project, which makes it difficult to compare or include as there may be many more changes than you actually want to add.
> 
> On the other hand, it is now only "+278/-3", which is much less than the previous "+3,652/-2,191". So I think the developer can work with that.

I am sorry for the confusion
I recognized yesterday that the implementation is already a at least SunSpec meter compatible implementation.
It looks like (at least) fronius inverters then only have additional checks for e.g. vendor to allow them beeing configured.

In a yet (in my repository) not commited change I created configuration for the necessary sunspec fields so it may be compatible with other inverters which do the same.

I will update this PR as soon as I have tested all the changes well enough.

bn-aklingler committed:
from Fronius-Simulation to SunSpec-OpenDTU compatibility.
add configuration for elements to equip with custom vendor strings.

bn-aklingler committed:
fix delayed starting of modbus server

AndreasBoehm commented:
There are still a lot of unrelated changes, please make sure that you base your branch off of the OpenDTU-OnBattery development branch.

> to the PR: the tree conflicting files webapp_dist/* need to be resolved on merge. other conflicts I could solve.
Please do not commit any file within `webapp_dist/`. This is done by the upstream project `OpenDTU` but within the OnBattery repo we don't do that, not sure why but i can imagine that its because of the conflicts that would arise constantly.

AndreasBoehm commented on include/Configuration.h:
```diff
@@ -76,6 +76,16 @@ struct CONFIG_T {
         bool Enabled;
     } Mdns;
 
+    struct {
```
Please also add `bool verboseLogging` to your struct and use that to enable/disable logging within your Modbus module instead of commenting the logging lines.

AndreasBoehm commented on platformio.ini:
```diff
@@ -45,6 +45,7 @@ lib_deps =
     nrf24/RF24 @ 1.4.9
     olikraus/U8g2 @ 2.35.19
     buelowp/sunset @ 1.1.7
+    https://github.com/emelianov/modbus-esp8266#master
```
Can we point to a specific commit or release-tag instead of a branch?
This might break builds at any point in the future without us knowing whats going on.

AloisKlingler replied:
currently not, as there is a change which is most probably needed in master only:
https://github.com/emelianov/modbus-esp8266/pull/292

as soon as there is a 4.1.2 (or whatever) released I think this one could be used.

AndreasBoehm replied:
If you specific the commit hash with `#a5872af7250e03e9a5cd578366b0c2cdc097fcc9` instead of `#master` you can make sure that the dependency won't get updated randomly with the next push to master

AndreasBoehm commented on src/ModbusDtu.cpp:
```diff
@@ -0,0 +1,230 @@
+                        value = (inv->Statistics()->getChannelFieldValue(TYPE_AC, CH0, FLD_PF));
+                        mb.Hreg(0x9cb9, hexbytes[1]);
+                        mb.Hreg(0x9cba, hexbytes[0]);
+                        // mb.Hreg(0x9cbb, 0);
+                        // mb.Hreg(0x9cbc, 0);
+                        // mb.Hreg(0x9cbd, 0);
+                        // mb.Hreg(0x9cbe, 0);
+                        // mb.Hreg(0x9cbf, 0);
+                        // mb.Hreg(0x9cc0, 0);
+                        // value = (inv->Statistics()->getChannelFieldValue(TYPE_AC, CH0, FLD_YT)*1000); //done above already!
+                        // if (value > _lasttotal) {
+                            // _lasttotal = value;
+                            // mb.Hreg(0x9cc1, hexbytes[1]); //done above already!
+                            // mb.Hreg(0x9cc2, hexbytes[0]); //done above already!
+                        // }
```
Please remove commented lines if they are not needed

AndreasBoehm commented on src/ModbusDtu.cpp:
```diff
@@ -0,0 +1,230 @@
+
+ModbusDtuClass ModbusDtu;
+
+ModbusDtuClass::ModbusDtuClass()
```
Wouldn't it make more sense if we call it `ModbusSunspec` or `SunspecModbus` to explain what this is actually doing?

Do you have any documentation about the protocol that youa re implementing here?

AloisKlingler replied:
this implementation is (maybe only partially) modbus-sunspec compatibility to simulate a sunspec energy-meter.

as far as I could google, this should match to "sunspec meter" registers in float mode.
I could not check if it is the complete protocol or only partial.
at least, if I check against fronius documentation, it is complete.

AloisKlingler commented:
@AndreasBoehm @schlimmchen @stefan123t
if I would change the above things, will the PR be merged?
what I can tell, the last released version from https://github.com/AloisKlingler/OpenDTU-FroniusSM-MB/releases/tag/v24.8.5-0.3.1 (the one with this PR) is running absolutely stable for me with one inverter. also about 10-20 users from photovoltaikforum.com use it and do not report issues.

I am not sure when I find time to do the changes requested above. my issue is, that adapting / merging to current OpenDTU versions (or also OpenDTU-OnBattery) is very time consuming caused by my knowledge.

it looks like there are some enhancement requests regarding ModbusTCP, I am happy if I can help, but I maybe need your help too if it is really wanted to be merged.

Thanks.

Best regards
Alois

AndreasBoehm commented on src/ModbusDtu.cpp:
```diff
@@ -0,0 +1,230 @@
+            mb.Hreg(0x9cc2, hexbytes[0]);
+        }
+
+        if (Hoymiles.getNumInverters() == 1) {
```
what if a user has configured multiple inverters?
Now that the DPL supports multiple inverters we shall not limit this feature a single inverter only.

AloisKlingler replied:
with multiple inverters the modbus data for total power (40097 and 40098) and yieldtotal (40129 and 40130) is filled.
those two values are the important one which are accumulateable easily (like already done in the "Total Yield Total" and "Total Power" in webgui)

other data (like frequency, voltage, etc.) is not filled, as we do not know from which source to take, and accumulating / calculating averages makes no sense at all.
this data is not mandatory (at least not for fronius inverters) to accept the ModbusTCP as smartmeter.

AndreasBoehm commented:
Let's talk about more about what this PR or your feature actually does, at least how i understood it. Please correct me if any of my assumptions are wrong.

- Its a ModbusTCP server that will be polled by other hardware to get data
- For this ModbusTCP server you implemented registers according to the 'SunSpec'-protocol

Changes that i would suggest
- it should have its own settings page as i don't see that its related to 'Network'
- classes, structs, fields need clear names, keeping ModbusTCPServer in mind
- cleanup the code, add more comments to make it easier to read and undrstand
- link a sunspec protocol documentation

I am wondering if this would be something that should be in the upstream project as its not related to having a battery at all?
https://github.com/tbnobody/OpenDTU

AloisKlingler commented:
> Let's talk about more about what this PR or your feature actually does, at least how i understood it. Please correct me if any of my assumptions are wrong.
> 
> * Its a ModbusTCP server that will be polled by other hardware to get data
> * For this ModbusTCP server you implemented registers according to the 'SunSpec'-protocol

exactly.

> 
> Changes that i would suggest
> 
> * it should have its own settings page as i don't see that its related to 'Network'
> * classes, structs, fields need clear names, keeping ModbusTCPServer in mind
> * cleanup the code, add more comments to make it easier to read and undrstand
> * link a sunspec protocol documentation
> 

I copied existing elements with great difficulty to have the here available settings. I am sure I cannot manage to create its own page

renaming I can do, only the here mentioned? https://github.com/hoylabs/OpenDTU-OnBattery/pull/1150#discussion_r1724430154

code cleanup I think I can not do satisfying enough. I tried to make the code already best readable as possible for me.

the sunspec protocol documentation looks to be liable to pay costs: https://sunspec.org/sunspec-modbus-specifications/
I had as reference fronius registers (https://www.fronius.com/de/solarenergie/installateure-partner/technische-daten/alle-produkte/anlagen-monitoring/offene-schnittstellen/modbus-tcp) and also kostal registers (https://optosolar.se/produktfiler/modbusprotkollengelskakostalplenticoreplus.pdf)

> I am wondering if this would be something that should be in the upstream project as its not related to having a battery at all? https://github.com/tbnobody/OpenDTU

there a previous modbus implementation was asked to offer a PR here.

stefan123t commented:
@AloisKlingler there are three compatibel standards mentioned on the sunspec page you referred to:

* [IEEE 1547-2018](https://ieeexplore.ieee.org/document/8332112) - IEEE Standard for Interconnection and Interoperability of Distributed Energy Resources with Associated Electric Power Systems Interfaces (aka SunSpec)
* [IEEE 2030.5-2018](https://ieeexplore.ieee.org/document/8608044) - IEEE Standard for Smart Energy Profile Application Protocol
* [IEEE 1815-2012](https://ieeexplore.ieee.org/document/6327578) - IEEE Standard for Electric Power Systems Communications-Distributed Network Protocol (DNP3)

I have been able to download the [IEEE_1547-2018_SunSpec.pdf](https://github.com/user-attachments/files/18658171/IEEE_1547-2018_SunSpec.pdf) PDF specifications for Distributed Energy Resource (DER) interoperability using the protocols defined in IEEE 2030.5-2018, IEEE 1815-2012 and SunSpec ModBus over TCP/IP transport via Ethernet / RS-485 (only the latter) from the above site. The other two are behind a paywall, access to which might be available to scientific researchers.

I agree with you all that it would be worthwhile to include this in the upstream project as it enables a standard ModBus interface besides MQTT and Web API to share data from the Hoymiles DTUs in the OpenDTU project.
Unfortunately @tbnobody and @schlimmchen have shown no need to include this functionality neither here https://github.com/tbnobody/OpenDTU/pull/582 nor there https://github.com/tbnobody/OpenDTU/pull/1893 so it was tried to add it to OpenDTU-OnBattery as it may be more applicable to integration with other Distribute Energy Resources.

I hope this may be (re-)considered at a later stage, once the PR has been more refined to both projects coding standards.

tbnobody committed:
Added missing breaking change to README.md

vaterlangen committed:
Fix: Corrected typo in string

schlimmchen committed:
Fix: typo in logging tag of PinMappingClass

tbnobody committed:
Fix: Dont allow HA base topic without trailing slash

schlimmchen committed:
MessageOutput: remove redundant #include

schlimmchen committed:
do not derive MessageOutput from Print

force us to use ESP_LOGx macros to log messages. print*() functions are
no longer available through MessageOutput.

schlimmchen committed:
remove downstream-specific task translations

these tasks are not part of OpenDTU, but OpenDTU-OnBattery.

schlimmchen committed:
webapp: translate logging modules in admin view

schlimmchen committed:
webapp: apply solarized-dark colors to web console

AndreasBoehm committed:
webapp: fix: return log level for more than 5 modules

tbnobody committed:
webapp: Update dependencies

schlimmchen committed:
MessageOutput does not use LOG macros by itself

schlimmchen committed:
Fix: make logging more leightweight using large shared buffer

LoQue90 commented:
with victron adding the posibility to send a limit for pv inverters over modbus to any inverter that follows the sunspec protocol, adding this to openDTUonB would be great.
https://community.victronenergy.com/t/help-needed-sma-solar-edge-kostal-enphase-and-others/33869

someone already did add this question to https://github.com/tbnobody/OpenDTU/pull/1893

i do have not enough experience in coding to help build this, but i would like to help as much as i can.
I´ve running 4 hm-800 with openDTUonB and a ac coupled victron ess.

tbnobody committed:
Add serial number prefix 114a to HMS_2CH

Fixes #2741

schlimmchen committed:
syslog: implement rate limiting

enabling a lot of logging messages, e.g., by setting verbose as the
default level, can make the system unresponsive if syslogging is
enabled. to protect against this, a rate limiting algorithm is
implemented in the syslogger.

tbnobody committed:
webapp: Update dependencies

This also increases the used node version from 22 to 24

LoQue90 commented:
@AndreasBoehm i thought I don't add another FR because the first underlying need Modbus integration.
But I will pull together what is needed and open up a FR.
As far as I understand how opendtu and openDTUonB depending on each other, the Modbus part should be integrated in opendtu first?
Thanks for your work!

AndreasBoehm commented:
This PR does integrate modbus, read only, but as discussed earlier in this PR we need to do a lot of cleanup to get this merged -> it probably makes more sense to start from scratch for this feature.

OpenDTU-OnBattery is based on OpenDTU, but not depending on it. If we want to integrate a feature we can do so, without the need to have this in OpenDTU first.

tbnobody committed:
Merge branch 'pr2713' into dev

tbnobody committed:
Merge branch 'pr2675' into dev

tbnobody committed:
Upgrade olikraus/U8g2 from 2.36.5 to 2.36.12

tbnobody committed:
Update bblanchon/ArduinoJson from 7.4.1 to 7.4.2

tbnobody committed:
Upgrade ESP32Async/ESPAsyncWebServer from 3.7.7 to 3.7.9

tbnobody committed:
Use fixed version of AsyncTCP

tbnobody committed:
webapp: Update node dependency

tbnobody committed:
Add Yarn Prettier badge to readme

schlimmchen committed:
logging: centralize rate limiting

schlimmchen committed:
logging: unlock while processing buffer

allows high priority tasks to log another message while the buffer is
being processed.

tbnobody committed:
BREAKING CHANGE: WebAPI endpoint `/api/limit/config` requires different parameters

Due to the fact that different inverters require different type identifiers in the backend the `limit_type` parameter was changed to numbers from 0 to 3.
* AbsolutNonPersistent = 0
* RelativNonPersistent = 1
* AbsolutPersistent = 2
* RelativPersistent = 3

tbnobody committed:
webapp: Update dependencies

tbnobody committed:
Upgrade ESP32Async/ESPAsyncWebServer from 3.7.9 to 3.7.10

schlimmchen committed:
update discord link to point to expected channel

tbnobody committed:
github actions: Upgrade actions/checkout to v5

tbnobody committed:
Upgrade ESP32Async/ESPAsyncWebServer from 3.7.10 to 3.8.0

tbnobody committed:
Remove comment which doesn't make sense at this posistion

tbnobody committed:
webapp: Updated to latest sass version

Including several silenceDeprecations settings as shown in https://github.com/twbs/examples/blob/main/vite/vite.config.js

tbnobody committed:
Feature: Allow setting of power limit on new firmware versions for HMS inverters

See #2784

tbnobody committed:
webapp: Simplify eventlog display

tbnobody committed:
webapp: Simplify line class detection in console view

tbnobody committed:
webapp: Make compiler happy

tbnobody committed:
webapp: Make compiler happy

tbnobody committed:
webapp: Doublecheck upload files bounds

tbnobody committed:
webapp: Use correct method scope

tbnobody committed:
webapp: Doublecheck for undefined behavior

tbnobody committed:
webapp: Fix lint error

tbnobody committed:
webapp: Make compiler happy

tbnobody committed:
webapp: Update dependencies

tbnobody committed:
webapp: Comment bootstrap workaround

tbnobody committed:
Use pre-defined literals in webapi

tbnobody committed:
webapp: Simplify authentication code

tbnobody committed:
webapp: Configure additional compression

tbnobody committed:
webapp: Fix exception on file upload select

tbnobody committed:
webapp: Minify json configs before upload

Leads to quicker parsing and downloading (in case of language files)

tbnobody committed:
webapp: Remove unused variable

tbnobody committed:
Added hint for breaking change

tbnobody committed:
webapp: add app.js.gz

tbnobody committed:
fix URL in greek language pack

tbnobody committed:
Updated discord link to delegate user directly into the right channel

tbnobody committed:
webapp: Update dependencies

mathieucarbou committed:
ESP32Async/ESPAsyncWebServer @ 3.9.2

tbnobody committed:
webapp: Prevent compile errors on newer tsc

tbnobody committed:
webapp: Update dependencies

tbnobody committed:
Merge branch 'pr2905' into dev

tbnobody committed:
Upgrade ESP32Async/ESPAsyncWebServer from 3.9.2 to 3.9.3

tbnobody committed:
Upgrade espMqttClient from 1.7.0 to 1.7.2

tbnobody committed:
Upgrade olikraus/U8g2 from 2.36.12 to 2.36.15

psiegert committed:
Add support for HMS-2000D-4T inverters with serial prefix 0x1166

Fixes recognition of Thai/Asian HMS-2000D-4T inverters that use serial
numbers starting with 1166. Without this fix, these inverters are not
properly recognized and the wrong RF module (not CMT2300A) is selected.

Tested with HMS-2000D-4T serial starting with 1166.

tidabek committed:
Update pl.lang.json

A few corrections to the Polish translation

tbnobody committed:
Merge branch 'pr2961' into dev

tbnobody committed:
Merge branch 'pr2953' into dev

tbnobody committed:
Added Serial Prefix to readme

tbnobody committed:
webapp: Update dependencies

tbnobody committed:
Upgrade ESP32Async/ESPAsyncWebServer from 3.9.3 to 3.9.6

tbnobody committed:
Upgrade olikraus/U8g2 from 2.36.15 to 2.36.17

tbnobody committed:
Rework NTP handling

* Use mathieucarbou/MycilaNTP which make the same as the currenet code with further improvements
* Remove zones.json file as it is already included in the NTP library now
* Remove the "Timezone" property from config as we will only save the time zone name in future (instead of spec +  name)
* When changing the time zone, the device local time will be updated immidiatly in the web ui

tbnobody committed:
webapp: Update dependencies

tbnobody committed:
Upgrade ESP32Async/ESPAsyncWebServer from 3.9.6 to 3.10.0

tbnobody committed:
Allow log level configuration for async websockets and async tcp stack

tbnobody committed:
Fix: Websockets did not work on latest iOS versions

On iOS and iPadOS there seem to be some very strange timing related issues. This new connect/reconnect handling should fix this.

tbnobody committed:
Upgrade ESP32Async/ESPAsyncWebServer from 3.10.0 to 3.10.3

tbnobody committed:
webapp: Update dependencies

tbnobody committed:
webapp: Change build config to match current vue pre-definitions

tbnobody committed:
Upgrade bblanchon/ArduinoJson from 7.4.2 to 7.4.3

tbnobody committed:
Upgrade olikraus/U8g2 from 2.36.17 to 2.36.18

tbnobody committed:
Upgrade nrf24/RF24 from 1.4.11 to 1.5.0

tbnobody committed:
Upgrade arkhipenko/TaskScheduler from 3.8.5 to 4.0.4

The _TASK_THREAD_SAFE is not required as all our tasks run in the same FreeRTOS thread.

tbnobody committed:
webapp: NTP admin did not receive proper headers

tbnobody committed:
Upgrade actions/checkout from v5 to v6

tbnobody committed:
Upgrade actions/setup-node from v4 to v6

tbnobody committed:
Upgrade node to 24 in github actions

tbnobody committed:
Upgrade actions/setup-python from v5 to v6

tbnobody committed:
Upgrade actions/stale from v9 to v10

tbnobody committed:
Upgrade dessant/lock-threads from v5 to v6

tbnobody committed:
Upgrade actions/cache from v5 to v6

tbnobody committed:
Upgrade actions/upload-artifact from v4 to v7

tbnobody committed:
Upgrade mikepenz/release-changelog-builder-action from v5 to v6

tbnobody committed:
Upgrade actions/download-artifact from v4 to v8

tbnobody committed:
webapp: add app.js.gz

tbnobody committed:
Updated copyright information

tbnobody committed:
Adjusted copyright information

tbnobody committed:
webapp: Add response error handling to getTimezoneList()

pieterhollander committed:
Fix: WebSocket endpoints missing setAuthType for digest authentication

tbnobody committed:
Merge branch 'pr3008' into dev

tbnobody committed:
webapp: add app.js.gz

bn-aklingler committed:
Merge upstream/master: integrate 511 commits from tbnobody/OpenDTU

Preserves custom Modbus/SunSpec configuration while integrating:
- New Syslog support
- Updated TaskScheduler (4.0.4) and MycilaNTP (6.0.5)
- 200+ file updates: new components, refactors, inverter support
- Webapp improvements and TypeScript updates
- Build artifact updates

Resolved conflicts in configuration structures, network API, and locale files.

bn-aklingler committed:
Update webapp build artifacts

bn-aklingler committed:
Fix: Replace MessageOutput.printf with ESP_LOGI for upstream compatibility

bn-aklingler committed:
Update: Normalize line endings in webapp and build artifacts

bn-aklingler committed:
Fix: Replace C-style cast with static_cast for cpplint compliance
